### PR TITLE
Add puppet 4 to the travis matrix

### PR DESCRIPTION
**Pull request by ekohl**

### CODE REVIEW

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -2,7 +2,6 @@
 #   https://github.com/theforeman/foreman-installer-modulesync
 
 require 'puppetlabs_spec_helper/module_spec_helper'
-require 'lib/module_spec_helper'
```
Comment (domcleal) [inline spec/spec_helper.rb]: This doesn't look right.. causing broken tests?

Comment (ekohl) [inline spec/spec_helper.rb]: I was looking at our modules and I wonder if we should include that helper in modulesync.

Comment (domcleal) [inline spec/spec_helper.rb]: Yeah, I was thinking the same!

**Comment by ekohl** [comment]
It now fails in the report processor, but I don't know if that's test related or just a general incompatibility with puppet 4.

**Comment by domcleal** [comment]
General incompatibility, but easily fixed it seems.  It looks like it unlocks a whole bunch of other failures, which we should see from Travis soon.

**Comment by ekohl** [comment]
Much of the failures are now due to a missing ```$::rubyversion``` fact in our tests.

**Comment by ekohl** [comment]
Rebased, now with more rspec-puppet-facts usage. It now fails on the foreman_report_processor_spec but haven't had time to look into it.

**Comment by domcleal** [comment]
> It now fails on the foreman_report_processor_spec but haven't had time to look into it.

I think our deserialisation of old reports into Puppet 4 is failing, probably due to internal class changes.  We might have to recreate serialised reports showing the different error conditions and separate them by version (2/3 vs 4).

There's a failure remaining in foreman_config_spec where apache::vhost's failing with a mix of non-IP and IP vhosts.  This used to work before r-p-f because $::interfaces was set to '' and https://github.com/theforeman/puppet-foreman/blob/4.0.0/manifests/config/passenger.pp#L89-L90 would ignore it.  Not quite sure how this is meant to work, tbh.

**Comment by ekohl** [comment]
> It now fails on the foreman_report_processor_spec but haven't had time to look into it.
> 
> I think our deserialisation of old reports into Puppet 4 is failing, probably due to internal class changes. We might have to recreate serialised reports showing the different error conditions and separate them by version (2/3 vs 4).

Makes sense. A shortcut would be to throw an error if we detect Puppet 4 to say it's not yet supported. Benefit would be that we test most of the code with Puppet 4 already.

> There's a failure remaining in foreman_config_spec where apache::vhost's failing with a mix of non-IP and IP vhosts. This used to work before r-p-f because $::interfaces was set to '' and https://github.com/theforeman/puppet-foreman/blob/4.0.0/manifests/config/passenger.pp#L89-L90 would ignore it. Not quite sure how this is meant to work, tbh.

I noticed that too and will have a look at it. I think it's time to revisit https://github.com/theforeman/puppet-foreman/pull/255

**Comment by mmoll** [comment]
As this already needs a serious rebase (and I failed to get the tests right), I propose to...
* open a new PR to get the compatible parts of this PR into master
* work on #255 to get it into master
* continue here

**Comment by ekohl** [comment]
> As this already needs a serious rebase (and I failed to get the tests right), I propose to...
> 
>     open a new PR to get the compatible parts of this PR into master

I'll do some cherry picking.

>     work on #255 to get it into master

That may not be needed, but probably not a bad thing in general.

**Comment by mmoll** [comment]
what's the status here?

**Comment by mmoll** [comment]
rebased, using #368 and the diff I posted there.

**Comment by mmoll** [comment]
added `foreman_url` also in spec/defines/foreman_config_passenger_fragment_spec.rb, now tests should become green

**Comment by domcleal** [comment]
Looks fine to me I think.  My only question is - what happened to the earlier test failures (report serialisation and vhost related)?  Something changed externally to fix it?

**Comment by mmoll** [comment]
I also asked that myself, the only thing I can imagine is that e.g. the interfaces/ip facts are now populated correctly via updated rspec-puppet-facts, but unsure if that's the real explanation.

**Comment by domcleal** [comment]
Rebased it on top of the rspec-puppet-facts merge to master.

File: .sync.yml [hunk 2]
```diff
@@ -16,6 +16,7 @@ Rakefile:
     - manifests/plugin/*.pp
 spec/spec_helper.rb:
   requires:
+    - lib/module_spec_helper
```
Comment (ekohl) [inline .sync.yml]: I think this can be dropped by now since we merged those into ```spec_helper.rb```.

Comment (domcleal) [inline .sync.yml]: Thanks, removed that commit.

**Comment by mmoll** [comment]
re-added 465f736

**Comment by ekohl** [comment]
Amended https://github.com/theforeman/puppet-foreman/commit/58a2802af4d6fd3dcc6da4aeeea0f0dfbcc1768f with spec tests for the passenger fragment defines.

**Comment by mmoll** [comment]
this is looking good :+1:

**Comment by domcleal** [comment]
Thanks @ekohl and @mmoll, merged!

**Pull request merged**